Fix Doctests for 0.5-dev

omus opened the pull request:
Updates the doctests to be compliant with Julia 0.5-dev. This PR is based off of the work done in #15753.

Doctests currently have 14 remaining failures that can be categorized as:
- ~~Regression for display of tuples of arrays (#17087)~~
- ~~Regression for display of `macroexpand` (#17090)~~
- ~~`dump` function no longer prints trailing newline (#17202 / [julia-dev forum post](https://groups.google.com/forum/#!topic/julia-dev/uHpfvBkClTc))~~
- Regression for `jl_` printing (#17107)

Requires that JuliaLang/JuliaDoc#23 be merged. (DONE)

vchuravy committed:
Fix doctest code-blocks in documentation

cb8aa0b - fix most doctests in the manual that changed due to backtraces.
cd3797e - fix stdlib doctests
1892f72 - fix doctests in devdocs
d9495c7 - resolve Warning mismatch by renaming functions
83ef854 - convert superflous doctest to code-block
78f6d70 - Fix example about typemax and typemin

          Due to the changes in 4706184 tuples are now printed compact. Thus to
          demonstrate the point in this example we have to execute the examples
          independently.
c03e6e0 - fixup linenumbers
35a31bb - start->first

Ref #15753

omus committed:
Corrected doctest exceptions

omus committed:
Doctest corrections related to changes to display

omus committed:
Limited number of + methods in doctest

omus committed:
Update documentation to use `view` instead `sub`

omus committed:
Update doctest for `expand`

omus committed:
Update doctest for `macroexpand`

omus committed:
Update doctest to work Nullable get

See JuliaLang/julia#16901

omus committed:
Update manual on method ambiguity on usage

When ambiguous methods now report an error on usage rather than
definition: see JuliaLang/julia#16125

omus committed:
Doctest fixes for dump display

omus committed:
Fix doctest for value types

omus committed:
Work around for doctest issue with empty line

Sphinx does not work well when docstring output contains an empty line.
Since `@code_warntype` has a empty line in the printed output this was
causing Sphinx to have incorrect expected data. As it turns out the
docstring example is only interested in the return type of the method
which meant we could truncate part of the output using ... and avoid the
empty line issue.

omus committed:
Update Stack Traces docs to match implementation

The Stack Traces section of the manual was very outdated.

vchuravy commented:
Thanks for picking this up

tkelman commented on base/docs/helpdb/Base.jl:
```diff
@@ -5643,7 +5643,8 @@ julia> deleteat!([6, 5, 4, 3, 2, 1], 1:2:5)
 ERROR: ArgumentError: indices must be unique and sorted
- in deleteat! at array.jl:543
+ in deleteat!(::Array{Int64,1}, ::Tuple{Int64,Int64}) at ./array.jl:571
+ in eval(::Module, ::Any) at ./boot.jl:231...
```
same comment as before, doesn't the doctest system already do some pruning of backtraces and wouldn't this be exactly the kind of thing to add to it?

omus replied:
The doctest system was doing some pruning before but it seemed to only work for entire lines. The ellipsis works much better for us as sometimes we append extra information afterwards such as `(repeats 2 times)`:
```
ERROR: ArgumentError: indices must be unique and sorted
 in deleteat!(::Array{Int64,1}, ::Tuple{Int64,Int64}) at ./array.jl:571
 in eval(::Module, ::Any) at ./boot.jl:231 (repeats 2 times)
 in _start() at ./client.jl:359
```

tkelman replied:
makes sense, but the `eval(::Module, ::Any)` part isn't really useful information - would it work if we put the ellipsis at the end of the previous line?

omus replied:
I am ok with doing that change. I didn't do that as the original stack traces included `eval(::Module, ::Any)`. 

@tkelman just to confirm: You want me to removing the `eval(::Module, ::Any)` from all (almost all) stack traces in the manual?

tkelman replied:
What do you mean by "original stack traces" ? Like what you get at the REPL or from running this code in a script (they'll look slightly different)? I think we already have the docsystem cutting out the pieces that appear in every single backtrace, 0.5 just adds a new such line. Since it's mostly noise I think it would be better not to add it. But would like to hear other opinions on the matter too.

omus replied:
> What do you mean by "original stack traces" ?
I meant what existed in the manual prior to this PR

tkelman replied:
Before this PR I'm seeing a handful, but quite a bit fewer than this PR is adding:
```
$ git grep -n 'in eval'
base/test.jl:914: in eval(::Module, ::Any) at ./boot.jl:226
doc/manual/mathematical-operations.rst:388:     in eval at ./boot.jl:263
doc/manual/mathematical-operations.rst:396:     in eval at ./boot.jl:263
doc/manual/mathematical-operations.rst:401:     in eval at ./boot.jl:263
doc/manual/mathematical-operations.rst:416:     in eval at ./boot.jl:263
doc/stdlib/test.rst:283:        in eval(::Module, ::Any) at ./boot.jl:226
test/core.jl:336:@test_throws ErrorException eval(:(abstract Qux2_{T} <: Sup2b_{Qux2_{Int}, T})) # wrapped in eval to avoid #16793
```

omus replied:
I stand corrected. It turns out that ce62f34 added in the eval line which I've been ignoring in my diffs. I'll remove those lines and probably most of the ellipsis as well.

omus committed:
Remove eval from most stack traces

omus committed:
Correct some documentation changes

Corrections include:
- Remove addition of `eval(::Module, ::Any)` from most stacktraces
- Use doctest groups rather than appending numbers to function names

omus committed:
Improving methods(+) doctest

The `methods(+)` doctest seems to fail periodically due to the order
of the methods changing. It appears that blocks of these methods
typically stick together so I picked a nice reading block and attempted
to format with the least amount of ellipsis.

Here is an interesting failure which caused me to update the layout
of the doctest. Before this commit we used the following as a doctest:

```jldoctest
julia> methods(+)
 # 166 methods for generic function "+":
...
+(a::Float16, b::Float16) at float16.jl:136...
+(x::Float32, y::Float32) at float.jl:206...
+(x::Float64, y::Float64) at float.jl:207...
+(x::Bool, z::Complex{Bool}) at complex.jl:137...
+(x::Bool, y::Bool) at bool.jl:36...
+(x::Bool) at bool.jl:33...
+{T<:AbstractFloat}(x::Bool, y::T) at bool.jl:43...
+(x::Bool, z::Complex) at complex.jl:144...
+(x::Bool, A::AbstractArray{Bool,N<:Any}) at arraymath.jl:105...
+(x::Char, y::Integer) at char.jl:40
...
+(a, b, c, xs...) at operators.jl:119
```

The output was as follows:
```
 # 166 methods for generic function "+":
+(a::Float16, b::Float16) at float16.jl:136
+(x::Float32, y::Float32) at float.jl:206
+(x::Float64, y::Float64) at float.jl:207
+(x::Bool, z::Complex{Bool}) at complex.jl:137
+(x::Bool, y::Bool) at bool.jl:36
+(x::Bool) at bool.jl:33
+{T<:AbstractFloat}(x::Bool, y::T) at bool.jl:43
+(x::Bool, z::Complex) at complex.jl:144
+(x::Bool, A::AbstractArray{Bool,N<:Any}) at arraymath.jl:105
+(x::Char, y::Integer) at char.jl:40
...
+(a, b, c, xs...) at operators.jl:119
```

omus commented:
Things are in a much better state now. I think this can be merged before the linked issues are dealt with as we can always fix the related doctests later.

omus committed:
Doctest changes from #17113

omus committed:
Doctest changes relating to dump display

omus committed:
Doctest update to reflect latest code

omus committed:
Update JuliaDoc changeset

See JuliaLang/JuliaDoc#23

omus commented:
Just waiting on #17202. I've left the doctests for #17107 broken as that issue requires that a section of the manual be re-written.

omus committed:
Doctest fixes after rebase

omus commented:
Rebased. Fixed some line numbers that were changed. One new doctest failure has appeared with "verbose_fussy_sqrt" reporting a line number of 0 in a stack trace #17317

kshyatt commented:
Thanks for your patience and all your hard work on this, @omus.

omus commented:
No problem. I found the manual invaluable when learning Julia. Keeping things up to date is the least I can do.

omus committed:
Skip methods(+) doctest

This particular doctest ended up producing a lot of false positives as
it appears the ordering from method can change from commit to commit.
Since this doctest was failing so consistently I set this particular
doctest to be skipped.

Additionally I expanded the sample of + methods shown which was
previously more restricted in an attempt to make this doctest pass more
consistently.

omus commented:
Rebased again since #17202 has been merged. Doctests uncovered #17338 which make up 8/11 currently failing doctests. I additionally changed one of the doctests to be skipped (8683558) as it tended to give allot of false positives.

omus commented:
Another rebase since #17338 was fixed. Only three remaining failed doctests. Two require the documentation to be updated (#17107) and the other should be addressed by #17324.

tkelman commented on base/test.jl:
```diff
@@ -893,24 +893,19 @@ julia> typeof(f(1,2,3))
 Int64
 
 julia> @code_warntype f(1,2,3)
-Variables:
```
is Variables output no longer being included in `@code_warntype` intentional?

omus replied:
It was intentional. I can add it back if it is preferred. The blank line was being an issue and I felt that the variable listing didn't add anything to the example.

tkelman replied:
oh right I see the `...` now, I guess it's alright to leave it out

tkelman commented on doc/manual/control-flow.rst:
```diff
@@ -126,7 +126,7 @@ This means that new variables defined inside the ``ìf`` clauses can
 So, we could have defined the ``test`` function above as
 
-.. doctest::
+.. doctest:: leaky
```
where are these `leaky`, `bad-leaky`, etc documented?

tkelman commented on doc/manual/documentation.rst:
```diff
@@ -30,7 +30,7 @@ macros and pass them to the ``@doc`` macro just as well.
 Here is a more complex example, still using Markdown:
 
-.. doctest::
+.. code-block:: julia
```
I guess this wasn't outputting anything?

omus replied:
I changed this to a code block as Julia interprets `function bar(x, y) ...` as a function call to `bar` which raises an exception.

tkelman replied:
strange, I think it's been passing on release-0.4?

tkelman commented on doc/manual/metaprogramming.rst:
```diff
@@ -415,11 +418,16 @@ Basics
+        return :( println("Hello, world!") )
+    end
+
+.. testoutput::
```
so this runs as a doctest, checks the output matches but doesn't include the output in the rendered docs?

omus replied:
Correct. The original doctest was failing since creating the macro now was outputting:
```
@sayhello (macro with 1 method)
```
Using `testcode` and `testoutput` allowed me to keep the example tested while keeping the documentation the same. Note we did lose the `julia>` prompt.

tkelman commented on doc/manual/metaprogramming.rst:
```diff
@@ -620,11 +629,14 @@ Compare:
       args: Array{Any}((5,))
-        1: String "a ("
+        1: String
+          data: Array{UInt8}((3,)) UInt8[0x61,0x20,0x28]
```
hm I guess this was a little nicer in the old format? is this a change in `dump` ?

tkelman replied:
bisected to 3405d263194 so I guess it was intentional - the new output isn't quite as nice, but I guess it says "`dump` should not be overloaded" so I guess we're stuck with it

omus replied:
I agree the output isn't as nice. Would it be worth opening an issue to confirm the changes to String output was intentional?

tkelman replied:
might get closed with a "yes," maybe a comment on that dump pr to start with?

omus replied:
I decided to just make an issue as GitHub comments are hard to search for #17392

tkelman commented on doc/manual/methods.rst:
```diff
@@ -226,148 +235,29 @@ values is perhaps the single most powerful and central feature of the
 Julia language. Core operations typically have dozens of methods:
 
 .. doctest::
+   :options: +SKIP
```
I might take this out right before release and see how annoying it gets while backporting fixes

omus replied:
Fair enough. If we do want to require doctests to pass for PRs I think this particular doctest is going to be the main issue.

tkelman commented on doc/manual/methods.rst:
```diff
@@ -427,9 +312,9 @@ the intersection case:
-first, since otherwise the ambiguity exists, if transiently, until the
-more specific method is defined.
+It is recommended that the disambiguating method be defined first,
+since otherwise the ambiguity exists, if transiently, until the more
```
maybe rephrase "if transiently" to "albeit transiently" ?

omus commented:
Thanks for merging this @tkelman
